### DISCUSSION TITLE
Type file not being generated

### DISCUSSION THREAD

**Discussion by WonderPanda** [0no-co/gql.tada#49] [Q&A] [answered]
I've previously played around with `@0no-co/graphqlsp` on a project in order to get type checking through the TS language server, which worked great!

I'm really excited to play around with gql.tada now as the promise of type safe GQL without codegen has been something I've dreamed of for years! Incredible work you've been doing here :pray: 

Following through the installation instructions though I can't seem to get the types to be generated using `tadaOutputLocation`

```
    "plugins": [
      {
        "name": "@0no-co/graphqlsp",
        "schema": {
          "url": "http://localhost:41182/v1/graphql",
          "headers": { "x-hasura-admin-secret": "secret" }
        },
        "tadaOutputLocation": "./src/graphql-env.d.ts"
      }
    ]
```

My GraphQL server is definitely up and running and other than the `tadaOutputLocation` this configuration has worked for me.

I've tried fully closing and restarting VsCode multiple times but still no luck

Are there any diagnostics or anything that I can review to try and figure out why the types file isn't being generated?

**Comment by JoviDeCroock** [comment, marked as answer]
Have you ensured that your VSCode is using the workspace TSVersion? If so you can enable TSLogging and open the TSServer log, there you can look for `[GraphQLSP]` and it should give you sufficient debug information, if you don't find any occurrences then that means you aren't using the workspace TypeScript version in your VSCode.

**Reply by WonderPanda** [reply]
Interesting. I did have it using the TS workspace version as intructed from the installation docs but I eventually got it working by explicitly opening up a typescript file that fell inside the Tsconfig's `include`.

I'm using an NX monorepo so there are quite a few tsconfig files and I'm still trying to determine big picture the best way to integrate gql.tada there.

Are there any plans to allow generation of the introspected types through a CLI instead of making it the responsibliity of the compiler plugin?  I saw the other discussion about checking in this file but in the past I've always set it up so that these types of files are gitignored and then re-generated cleanly in CI.

Also do you have any thoughts or guidance on using gql.tada inside of a monorepo with many apps/projects? They all share our single common GraphQL backend at the moment and ideally I want to have a single introspected types file for the whole repo and not need to set this up on a per project basis.

Sorry for the random stream of consciousness! Let me know if you want me to create individual discussions for some of these questions. Again, appreciate all the incredible work you're doing for the GQL ecosystem!

**Reply by JoviDeCroock** [reply]
We do have plans for this https://github.com/0no-co/GraphQLSP/issues/212

**Reply by itsboogie** [reply]
@WonderPanda did you have any success setting this up in a NX Monorepo? I managed to get it to work if i set it up within a library but if i configure it in the root `tsconfig.base.json` then the `graphql-env.d.ts` is not being generated :/

**Reply by WonderPanda** [reply]
I got it working in an NX monorepo but ended up having to repeat the config across each lib's TsConfig file. I only generate the types in one library that is shared amongst the others though. All the others just declare the plugin in order to get autocomplete from the LSP

**Reply by colinwurtz** [reply]
@WonderPanda  or @itsboogie 
>  but ended up having to repeat the config across each lib's TsConfig file

Did you do this in your lib's `tsconfig.json` file or `tsconfig.lib.json` file? Or both?

**Reply by WonderPanda** [reply]
Just the tsconfig.lib.json. That's the one that actually specifies which TS files are being referenced through the `include` config:
![image](https://github.com/0no-co/gql.tada/assets/3631771/61d4e74f-088c-4806-9592-43049d84e501)

**Reply by WonderPanda** [reply]
I have a single library in my monorepo for which I actually configure the codegen tadaOutputLocation. I called that library `@my-workspace/gql.tada` and it exports the `graphql` function to be used by all other libraries.

The other libraries do need to configure the plugin in there `tsconfig.lib.json` but they do not need to specificy a tada output location or do any codegen. It is purely to get integration with the Typescript Language Server for autocomplete when authoring GQL.

**Reply by anirudhsama** [reply]
Hi @WonderPanda 

Does auto complete work when you configure it this way? I'm getting errors and warnings, but none of the auto complete features seem to be working. I'm assuming the `graphql.env.d.ts` file is saved in the `@my-workspace/gql.tada` package?

**Reply by anirudhsama** [reply]
Hey, got it working, it was some configuration issue on my end.

**Reply by itsboogie** [reply]
so just to be clear, all you include in the consuming libraries `tsconfig.lib.json` is: 

```
    "plugins": [
      {
        "name": "@0no-co/graphqlsp"
      }
    ]
```

because this is not working for me (auto complete works in the "providing" library [`@my-workspace/gql.tada`] but not in consuming libraries that only have the snippet above in their tsconfig :/ 

strangely enough the resulting types are correct, so i can technically use it and be typesafe, im just not getting autocomplete :sweat_smile:

**Reply by kitten** [reply]
@itsboogie The config you've got in `tsconfig` is basically there to tell GraphQLSP how to interpret the queries (besides the CLI also using it)

So if you don't hoist the config to a `tsconfig.json` you extend from or duplicate it, then GraphQLSP won't (for example) have any information about your schema to actually function outside of your subpackage 👀

**Comment by igortas** [comment]
I'm having similar problem in monorepo, ```.env.d.ts``` file is not generated.

```
{
  "typescript.tsdk": "node_modules/typescript/lib",
  "typescript.enablePromptUseWorkspaceTsdk": true
}
```

```
    "plugins": [
      {
        "name": "@0no-co/graphqlsp",
        "schema": "./schema.graphql",
        "tadaOutputLocation": "./src/graphql-env.d.ts"
      }
    ]
```

**Reply by kitten** [reply]
have you run `gql.tada doctor` or seen if the file generates with `gql.tada generate output`?

**Reply by Mingyang-Li** [reply]
Same, I'm also unable to generate `graphql-env.d.ts` from my monorepo. Where and how do you run `gql.tada doctor` ? @kitten 

I tried `npx gql.tada doctor` and `npx gql.tada generate output` in the monorepo root folder and the project folder, doesn't work

**Reply by kitten** [reply]
@Mingyang-Li: The doctor command should give you an indication of what's going on. Generate output also should tell you what's going on if it's not generating ab output file

**Reply by Mingyang-Li** [reply]
@kitten I tried the following commands, none of them helped. I kept getting the same error saying "Windows cannot open this type of file (.tada)", do you have any ideas how to resolve this?

Assume my frontend folder name is called `frontend` and I'm running the commands in the project folder of my monorepo (not root level). I've installed both the package itself as well as the LSP plugin
cmds I tried:
1. `npx gql.tada init ./frontend`
2. `pnpx gql.tada init ./frontend`
3. `npm gql.tada init ./frontend`
4. `gql.tada generate output`
5. `pnpm gql.tada init`
6. `npx gql.tada generate output`

**Reply by JoviDeCroock** [reply]
Try npx gql-tada, windows might have some weird quirks with the dot notation

**Reply by Mingyang-Li** [reply]
Thanks, `npx gql-tada generate output` works - I got my `graphql-env.d.ts` now.

I'm still not getting autocompletion as I type my queries in the `graphql()` function, nor an I getting syntax highlighting. Have I forgot to install any vscode plugins to enable this?

**Reply by kitten** [reply]
The plugin isn't loaded in your editor. If you have VSCode, make sure to follow the instructions in the note here: https://gql-tada.0no.co/get-started/installation#step-1-%E2%80%94-installing-packages

VSCode won't load any TypeScript plugins unless it's set to use "the workspace version of TypeScript" (fancy phrase they use to say, the local TypeScript installation in your `node_modules` folder)

Without the `tsconfig.json` being set up (which `gql-tada doctor` can check for you) and without VSCode (or another editor with TypeScript support) having loaded the plugin, no auto-completions, diagnostics, or hover info will show up for GraphQL strings.

**Reply by Mingyang-Li** [reply]
I did follow the steps and ran `npx gql-tada doctor`, here's the output:
```
C:\Users\FOLDER\my_monorepo\packages\frontend>npx gql-tada doctor
┏  Doctor 
┃ Detects problems with your setup
┃
┣ ✔ Checking TypeScript version
┃
┣ ✔ Checking installed dependencies
┃
┣ ✔ Checking tsconfig.json
┃
┗ ✔ Checking schema

 Done  You are all set and ready to go.
```

I my `.vscode/` folder (root folder, not the client app folder), I have this in my `settings.json`:
```json
{
  "typescript.tsdk": "node_modules/typescript/lib",
  "typescript.enablePromptUseWorkspaceTsdk": true
}
```

My `/packages/frontend/tsconfig.json`:
```json
{
  "extends": "tsconfig/base.json",
  "compilerOptions": {
    //vite config
    "jsx": "react-jsx",
    "allowSyntheticDefaultImports": true,
    "isolatedModules": true,
    "module": "ES2022",
    "moduleResolution": "Node",

    "sourceMap": true,
    "inlineSources": true,

    // gql.tada config
    "plugins": [
      {
        "name": "@0no-co/graphqlsp",
        "schema": "../backend/src/schema.graphql",
        "tadaOutputLocation": "./src/graphql-env.d.ts"
      }
    ],

    // Standard project config
    "declaration": false,
    "baseUrl": ".",
    "paths": {
      "@/*": ["./src/*"]
    }
  }
}
```

What do you see I'm doing wrong?

**Reply by kitten** [reply]
@Mingyang-Li: Most likely, seeing that only `/packages/frontend/tsconfig.json` has been edited, could it be that you're opening your editor in the root of your workspace and only the root `tsconfig.json` is being picked up by your editor?

**Reply by Mingyang-Li** [reply]
Potentially... lemme try only opening the project folder in my editor and see if autocompletion comes up.

**Reply by Mingyang-Li** [reply]
Here' s my root `tsconfig.json`:
```json
{
  "compilerOptions": {
    // gql.tada config
    "plugins": [
      {
        "name": "@0no-co/graphqlsp",
        "schema": "./packages/backend/src/schema.graphql",
        "tadaOutputLocation": "./packages/frontend/src/graphql-env.d.ts"
      }
    ],
  }
}
```

I've kept my frontend tsconfig file as it was.

No autocompletion.

I also tried only opening the frontend project folder just to see if my editor picks up the frontend tsconfig file, it didn't. I'm still not getting syntax highlighting nor am I getting autocomplete
